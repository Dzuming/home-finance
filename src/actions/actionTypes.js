export const OPEN_SIDENAV = 'OPEN_SIDENAV';
export const REQUEST_LOGIN = 'REQUEST_LOGIN';
export const SUCCESS_LOGIN = 'SUCCESS_LOGIN';
export const AUTH_USER = 'AUTH_USER';
export const LOGIN_FAILURE = 'LOGIN_FAILURE';
export const REQUEST_LOGOUT = 'REQUEST_LOGOUT';
export const SUCCESS_LOGOUT = 'SUCCESS_LOGOUT';
export const REQUEST_USER = 'SUCCESS_LOGOUT';
export const GET_USER = 'SET_USER';
export const SET_USER = 'SET_USER';
export const REMOVE_SPENDING = 'REMOVE_SPENDING';
export const EDIT_SPENDING = 'EDIT_SPENDING';
export const SET_DATE = 'SET_DATE';
export const API_REQUEST = 'API_REQUEST';
export const SET_CATEGORIES = 'SET_CATEGORIES';
export const SET_SPENDING = 'SET_SPENDING';
export const API_REQUEST_POST = 'API_REQUEST_POST';
export const API_REQUEST_DELETE = 'API_REQUEST_DELETE';
export const API_REQUEST_PUT = 'API_REQUEST_PUT';
export const ADD_SPENDING = 'ADD_SPENDING';
