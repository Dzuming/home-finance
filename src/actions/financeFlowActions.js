import * as types from './actionTypes';
import env from '../../environments/config';
import { financeFlowSpendingToServerMapper, financeFlowSpendingToTableMapper } from '../helpers/Mappers';

export const fetchCategories = () => ({
  type: types.API_REQUEST,
  payload:
    {
      url: `${env.api_url}/api/categories`,
      success: setCategories
    }
});

export const setCategories = (data) => ({
  type: types.SET_CATEGORIES,
  payload: data
});

export const fetchSpending = () => (dispatch, getState) => {
  const {id} = getState().user;
  const {selectedDate} = getState().financeFlow;
  return dispatch({
    type: types.API_REQUEST,
    payload:
      {
        url: `${env.api_url}/api/spending/${id}/${selectedDate}`,
        success: setSpending
      }
  });
};

export const setSpending = (data) => ({
  type: types.SET_SPENDING,
  payload: data
});

export const createSpending = (data) => ({
  type: types.API_REQUEST_POST,
  payload:
    {
      url: `${env.api_url}/api/spending`,
      success: addSpending,
      data
    }
});

export const addSpending = ({spending, message}) => ({
  type: types.ADD_SPENDING,
  payload: {
    spending,
    message
  }
});

export const deleteSpending = (id) => ({
  type: types.API_REQUEST_DELETE,
  payload:
    {
      url: `${env.api_url}/api/spending/${id}`,
      success: removeSpending,
      id
    }
});

export const removeSpending = (message, id) => ({
  type: types.REMOVE_SPENDING,
  payload: {
    message,
    id
  }
});

export const putSpending = (data) => ({
  type: types.API_REQUEST_PUT,
  payload:
    {
      url: `${env.api_url}/api/spending/${data.id}`,
      success: editSpending,
      id: data.id,
      items: financeFlowSpendingToServerMapper(data.items)
    }
});

export const editSpending = (message, id, items) => ({
  type: types.EDIT_SPENDING,
  payload: {
    message,
    id,
    items: financeFlowSpendingToTableMapper (items.data)
  }
});

// export function setDate (date) {
//   return {type: types.SET_DATE, date};
// }
