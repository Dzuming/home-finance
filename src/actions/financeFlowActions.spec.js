import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import * as actions from './financeFlowActions';
import * as types from './actionTypes';
import fetchMock from 'fetch-mock';
import expect from 'expect';
import env from '../../environments/config';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('login actions', () => {
  afterEach(() => {
    fetchMock.reset();
    fetchMock.restore();
  });

  it('creates SUCCESS_FINANCE_FLOW when fetchFinanceFlow has been done', () => {
    const userId = 1;
    const selectedDate = '2018-01';
    fetchMock
      .getOnce(`${env.api_url}/api/spending/${userId}/${selectedDate}`, {
        body: {description: '12345', value: 'John Doe', category: {id: 1}},
        headers: {'content-type': 'application/json'}
      });
    const expectedActions = [
      {type: 'SUCCESS_FINANCE_FLOW', data: {description: '12345', value: 'John Doe', category: {id: 1}}},
    ];
    const store = mockStore({user: {id: 1}, financeFlow: {selectedDate: '2018-01'}});
    return store.dispatch(actions.getFinanceFlow()).then(() => {
      // return of async actions
      expect(store.getActions()).toEqual(expectedActions);
    });
  });
});
