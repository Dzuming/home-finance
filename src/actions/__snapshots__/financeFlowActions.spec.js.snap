// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`finance flow actions should create SET_CATEGORIES action with snapshot 1`] = `
Object {
  "payload": Object {
    "id": 1,
    "name": "jedzenie",
  },
  "type": "SET_CATEGORIES",
}
`;

exports[`finance flow actions should create addSpending action with snapshot 1`] = `
Object {
  "payload": Object {
    "message": undefined,
    "spending": undefined,
  },
  "type": "ADD_SPENDING",
}
`;

exports[`finance flow actions should create editSpending action with snapshot 1`] = `
Object {
  "payload": Object {
    "id": 1,
    "items": Object {
      "category": "jedzenie",
      "dateCreated": "2018-01-01",
      "description": undefined,
      "id": 1,
      "spending": 300,
    },
    "message": "test",
  },
  "type": "EDIT_SPENDING",
}
`;

exports[`finance flow actions should create removeSpending action with snapshot 1`] = `
Object {
  "payload": Object {
    "id": undefined,
    "message": Array [
      Object {
        "id": 1,
      },
    ],
  },
  "type": "REMOVE_SPENDING",
}
`;

exports[`finance flow actions should create setSpending action with snapshot 1`] = `
Object {
  "payload": Array [
    Object {
      "category": Object {
        "id": 1,
        "name": "jedzenie",
      },
      "created_at": "2018-01-01 12:12:12",
      "id": 1,
      "name": "Zakupy",
      "user_id": 1,
      "value": 300,
    },
  ],
  "type": "SET_SPENDING",
}
`;
